### PR TITLE
⚠️ Prototype: Including Blurhash support for gatsby-image 🚀

### LINKED ISSUE #15563: Try adding support for https://blurha.sh/
I ran across https://blurha.sh/ recently (via @jamo) which looks like an excellent blur-up — better than our existing method of relying on the browser to upscale a 20px image.

Would be interesting to try adding support for this to gatsby-plugin-sharp/gatsby-transformer-sharp/gatsby-image.

This would be an extra field on the image type that does the pre-processing. The library also includes a JS client library ([1.6kb](https://bundlephobia.com/result?p=blurhash@1.1.0)) that we'd include with gatsby-image — probably as an optional export e.g. `import Img from "gatsby-image/with-blurha"` or we could document how to pre-process the generated string to convert it back to an image.

Would someone like to try putting together a prototype so we can see how this works for real? Perhaps add this as another tab to https://using-gatsby-image.gatsbyjs.org/

### PR DESCRIPTION
**Pull request by Darking360**
# ⚠️ Prototype: Including [Blurhash](https://github.com/woltapp/blurhash/tree/master/TypeScript) support 🚀

This PR is a prototype to include support for [Blurhash](https://blurha.sh/) inside `gatsby-image` which uses `gatsy-image-sharp` and `gatsby-transformer-sharp` in the process. This is a request by @KyleAMathews 🔥

## Description

Blurhash is an algorithm that takes an image and converts it into a string of 20 or 30 characters long. This is done with some canvas manipulation to get the image in `Uint8ClampedArray
` format to then get the blur hashed string representing the blurred image 🎉

## Implications of implementation

For this implementation I had to install 2 packages, one is [blurhash](https://github.com/woltapp/blurhash/tree/master/TypeScript) itself, which comes from the official Typescript implementation, and also add [Node Canvas](https://github.com/Automattic/node-canvas) to have a canvas API in Node, with these 2 packages it's possible to add a blurhash function inside the `generateBase64` function of `gatsby-plugin-sharp` to get a blurhashed image if options are supplied for that. I added `blurhashed`, `componentX` and `componentY` options inside the `gatsby-transformer-sharp` in `extend-node-types.js` file.

Some bundlephobia links for node canvas and blurhash.

- [Node Canvas](https://bundlephobia.com/result?p=canvas@2.5.0)
- [Blurhash](https://bundlephobia.com/result?p=blurhash@1.1.3)

Another thing is that I added the `node canvas` package alongside with `sharp` old package for handling image blur, the little issue is that Node gives a warning saying:

```
objc[19031]: Class GNotificationCenterDelegate is implemented in both /Users/miguel/contributed/gatsby-universal/node_modules/sharp/vendor/lib/libgio-2.0.0.dylib (0x10f5fb578) and /Users/miguel/contributed/gatsby-universal/node_modules/canvas/build/Release/libgio$
```

Maybe because both libraries uses `libgio`, but it's something to keep in mind 🤔 Another thing is that `Node Canvas` uses [Cairo Graphics](http://cairographics.org/) for its backend, but its binaries should be installed in the normal `install` process for `gatsby-plugin-sharp`.

## How does it work? 💥

As said, I added the functionality inside the `generateBase64` function, so any image using it should have access to it, but it works with conditionals from the options, I added those options to `fluid` image node type with the `blurhashed`, `componentX` and `componentY` which are set by default to `false`, `5` and `5` respectively, so for using it you should write your query like:

```
image {
  childImageSharp {
    fluid(maxHeight: 500, quality: 90, blurhashed: true) {
      ...GatsbyImageSharpFluid_withWebp
    }
  }
}
```

Specifying the property `blurhashed` as `true` should use the `fileBlurhashedToBase64` function inside `gatsby-plugin-sharp` to get the blurry image from `blurhash` library 🚀

## Related Issues

Fixes #15563

This prototype request was originally made by @KyleAMathews 

### First contribution 😅

It's my first time trying to contribute to Gatsby and checking the code of its core, or at least some little part and plugins as for images, so apologies in advanced for any bad code or practices, just blow my mind when I got into the guts of gatsby 🤩 Thanks a lot 🙏

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Yeah! Super excited to see this 😍

Could you add a demo of using this to https://using-gatsby-image.gatsbyjs.org in this PR and publish it somewhere so we can see this in action?

**Comment by Darking360** [comment]
Hey! Sure sure, I can add it ✍️ 

Actually I tested it with a starter project on Gatsby and with the `gatsby-dev` command. Let me see if I can have it something ready for tomorrow or maybe Friday EOD, I'll let you know 🚀 

Thank you ☺️

**Comment by Darking360** [comment]
Greetings  :sun_with_face: and once again sorry for the delay :sweat_smile: 

I just finished adding the example to the `using-gatsby-image` example page, I added 3 new unplash images to make it look different from the other `Blur Up` page, and add some information about it there too, I didn't change the links though :thinking: 

I had one problem running the project with `yarn develop` because of the image cache after the first run which was fine, saying:

```
Error: Invariant Violation: Encountered an error trying to infer a GraphQL type for: `localFile___NODE`. There is no corresponding node with the `id` field matching: "cf3e654595d61bf6b450a3f728b0687c".
```

I think it's because of the image cache with the remote files maybe :sweat_smile: still taking a look into that, but removing the `.cache` worked to generate the build and upload it to `now` so I attached the link here so you can check it out :rocket: 

https://public-kypgoyjipb.now.sh/blurhash/

*CC* @KyleAMathews

File: packages/gatsby-plugin-sharp/src/index.js [hunk 1]
```diff
@@ -137,13 +139,60 @@ function queueImageResizing({ file, args = {}, reporter }) {
...
+// Get blurhashed and transform to base64
+async function fileBlurhashedToBase64(
+  absolutePath,
+  width,
+  height,
+  componentX,
+  componentY
+) {
+  // Load image
+  const image = await loadImage(absolutePath, width, height)
```
Comment (KyleAMathews) [inline packages/gatsby-plugin-sharp/src/index.js]: would this be faster if we pre-processed the image to a smaller one e.g. 400px wide? Some images in people's site can be quite large. Not sure how well node-canvas scales

File: examples/using-gatsby-image/src/pages/blurhash.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      The default Blur Up technique uses progressive loading to make a fast,
+      visually pleasing experience without waiting for a full-resolution image
+      with a blank screen.
+    </p>
+    <h2>Blur Up with Blurhash library 🚀</h2>
+    <p>
+      In short, BlurHash takes an image, and gives you a short string (only
+      20-30 characters!). BlurHash applies a simple DCT transform to the image
+      data, keeping only the first few components, and then encodes these
+      components using a base 83 encoding, with a JSON, HTML and shell-safe
```
[the remark this answers is not in the document]
Comment (Mike-Dax) [reply examples/using-gatsby-image/src/pages/blurhash.js]: Yeah.
https://github.com/woltapp/blurhash/blob/master/Algorithm.md#base-83

**Comment by Darking360** [comment]
Greetings @KyleAMathews really glad that you liked it 🙏 👏 

For the first comment about the `base 83` yes it is correct, my bad for not including it since I didn't realize that it was a pretty custom encoding for the algorithm 😅  they explain it more in the algorithm [implementation here](https://github.com/woltapp/blurhash/blob/master/Algorithm.md) and talk more about the [base 83 here](https://github.com/woltapp/blurhash/blob/master/Algorithm.md#base-83) inside the same repository. I also found the `base 83` implementation inside the Typescript library used here to do the blurhash, so you can take a look at it [here](https://github.com/woltapp/blurhash/blob/master/TypeScript/src/base83.ts), I think it is a pretty cool implementation 🚀 I think these are good links to include in the documentation of the new page of using gatsby image, inside the `Blur Hash` link 😅 Let me know and I can add them 💪 

For the second question, I've been reading the implementation and most of it has some linear complexities like the `base 83` encoding for example, and others with `n*m`, being, for example, looping through the pixels width and height of the picture, or through the `componentX` and `componentY` provided to the function, that both values could possibly go from 1 to 9 🤔 but I'd need to run maybe some performance tests between the 2 `sharp` and `node canvas` to check how they deal with larger images, but in theory they should do less work if we take an image with less width or height 🤓

Let me know what to add or do from here, I think it'd be:

- [ ] Add useful links to the algorithm in the using gatsby image `Blur Hash` link.
- [ ] Compare `sharp` and `node canvas` implementations on `blur` to check performance on larger images (e.g. 1000px and beyond).

**Comment by Darking360** [comment]
@wardpeet Thank you so much! Glad you liked it 🚀 and just saw the changes, sure I can add those and test it again, I can have them for the weekend probably though 😅 

For the tests with `sharp` and `node canvas` I'll create a repo with some large images and run tests with both loading them and set tests with performance API and memory to check it out, as soon as I have both things I'll let you know with the results 💪 

Thanks again 💯

**Comment by wardpeet** [comment]
We resize images in parallel. Whenever a query is ran we resolve images and we bundle transforms of the same image together. Sharp is made to run parallel as it designed to do multi core processing. I have a feeling node-canvas isn't.

Another feeling I have is that nodejs doesn't have access to the heap usage as sharp & node-canvas are C libraries and live outside of nodejs.

I don't think these things are blocking as it's opt-in behaviour. So all good for me! We should do a benchmark with [blurhash wasm](https://github.com/fpapado/blurhash-rust-wasm) approach, see if it's an improvement we can do in a follow up pr.

Your benchmark was really awesome and handy, thank you so much! :heart: You deserve a 🥇.

I'll see what the rest of @gatsbyjs/core thinks to move forward with this.

**Comment by wardpeet** [comment]
We're a bit hesitant to merge it because of the node-gyp dependency. We want to make the dependency optional and if we fail to require it we should throw a structured log so people know what to do here.

i'm fine with taking up the work, if you want to do it I can put in a full explanation on how do drive this to the finish line.

What do you think?

**Comment by Darking360** [comment]
Greetings @wardpeet I see I see. In that scenario `node canvas` won't perform that better against `sharp`, I research a little bit to check if we could drop `node canvas` and only use `sharp` to read and get the bit array to use with `sharp`, and the problem is to convert a `Uint8ClampedArray` back to base64 😞 That's why the usage of `node canvas` and its support for canvas-like API 🤔 

About the rest of the work, I'm ok to take it and add the structured log, well if you're ok with it of course 😅 If you want to finish it for any reason I'm ok too, whatever that fits the most with your responsibilities and the Gatsby core too 💪 Let me know and I can do it 🚀 Thanks a lot for your work with this PR 🙏

**Comment by stefanoverna** [comment]
> This is even more exciting than I thought as I didn't realize that we could avoid loading their client library!

Wait @KyleAMathews... How it possible? :confused:

**Comment by wardpeet** [comment]
@darking360 is doing the blurring with node-canvas which makes it done on the server and not on the client.

Would it be possible to make a gatsby-plugin-blurhash like we have with sqip? @KyleAMathews do you want to integrate it in gatsby-plugin-sharp or does a separate plugin work fine too?

**Comment by KyleAMathews** [comment]
Separate plugin would be better to start with yeah

**Comment by wardpeet** [comment]
@Darking360 you already did so much! If you don't have the time to create a custom plugin out of this, feel free to tell us. I'm happy to copy and paste your code and give you credits for it if you don't have the time.

**Comment by Darking360** [comment]
Morning (from this part of the world hehe) @wardpeet :sun_with_face: 

Thank you! :pray: Actually I was waiting on instructions since I didn't know how to continue the work haha :sweat_smile: last time I read was to add a structured log since there were some conditional imports like the one for `canvas` for example :thinking: Of course I know that you have a bunch of work in the core team so it's ok.

As I mentioned before whatever works most for you, the idea is to ship this to check how it works inside Gatsby and give the users more options. If you want to continue it's perfectly fine, if not, I can help in whatever you tell me to, just let me know and I can jump in @wardpeet

**Comment by wardpeet** [comment]
Hey @Darking360, i'm so sorry about letting this slip. I hope you're doing well.

I would look at gatsby-transformer-sqip and copy the structure but use blurhash instead of sqip 😄.

Wdyt?

**Comment by Darking360** [comment]
Hey @wardpeet, how are you? No problem hehe I know there are tons of work :muscle: I hope you're doing well, and sure, taking the structure of `gatsby-transformer-sqip` sound good, let me know how to continue, thank you :pray:

**Comment by wardpeet** [comment]
Hey @Darking360. Oh i thought you could go forward with this PR. The idea is to create a `gatsby-transformer-blurhash` just like `gatsby-trasnformer-sqip`.

So instead of implementing this inside `gatsby-plugin-sharp` & `gatsby-transformer-sharp`. Later on, we can move it to gatsby-plugin-sharp when we have wasm in place.

I don't know if this is enough for you to go further? If not feel free to send me an email on ward@gatsbyjs.com and we can setup a call

**Comment by LekoArts** [comment]
Hey @Darking360 👋 Do you have time to set up a call with Ward?

**Comment by Darking360** [comment]
Greetings @LekoArts how are you?  👋 Sorry I didn't answer earlier. Sure! I can schedule a call with Ward. If it's ok, could it be after February 10th? Because I'll travel to Colombia to a conference tomorrow and I'll be back around February 10th, well, back to have free time I mean 😅 Let me know 💪

**Comment by m5r** [comment]
Hi, do you need help with this ?

**Comment by m5r** [comment]
As explained [in my comment](https://github.com/gatsbyjs/gatsby/issues/15563#issuecomment-609505949), i built a [gatsby transformer](https://github.com/m5r/gatsby-transformer-blurhash) to use blurhash

Do you want me to open a pull request and get a review going ?

**Pull request closed without merging**